### PR TITLE
Even Pods Spread - 3. Predicates Core

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/sig scheduling
/hold

/assign @bsalamat 
/cc @krmayankk 

**What this PR does / why we need it**:

This is the **3rd** PR of the "Even Pods Spread" KEP  implementation. After this PR, users can run workloads using "hard topologySpreadConstraints". (Preemption and Priority support PRs will sent out later)

- Define a new Predicate to supports API spec `whenUnsatisfiable: DoNotSchedule`
- Core Predicate logic
- unit tests on EvenPodsSpread predicate
- unit tests on genericScheduler.Schedule()

**Which issue(s) this PR fixes**:

Part of #77284.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

(will document all the changes in one place)

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/algorithm/predicates/metadata.go
Comment (krmayankk) [inline pkg/scheduler/algorithm/predicates/metadata.go]: is it necessary to use an existing struct from anti affinity ? I was thinking when we deprecate it, it might be easier to cleanup the code

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (krmayankk) [inline pkg/scheduler/algorithm/predicates/predicates.go]: why does this only get the hard topology constraints ?

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (krmayankk) [inline pkg/scheduler/algorithm/predicates/predicates.go]: when does this get populated ?

File: pkg/scheduler/util/utils.go
Comment (krmayankk) [inline pkg/scheduler/util/utils.go]: why do we care about the affinity fields in this PR ?I would try to keep this feature completely separate from the affinity feature

**Review by krmayankk — COMMENTED** [review]
i took a quick first pass over it. More comments will help understand the core logic

File: pkg/scheduler/algorithm/predicates/metadata.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: `topologyPairsMaps` is a common data structure, and is not limited to work for affinity only.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: As predicates only care about hard constraints :) In Priority logic, it will take care of soft constraints.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: As usual, at the beginning of Predicates phase.

File: pkg/scheduler/algorithm/predicates/metadata.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: Done.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (ahg-g) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Can you please add an example like the one you have in the KEP to document the reasoning behind this logic.

I would also compute the skew on a separate line above. 

```
skew := matchNum+selfMatchNum-minMatchNum
if skew > constraint.MaxSkew {
```

btw, is the casting to int32 actually necessary?

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (ahg-g) [inline pkg/scheduler/algorithm/predicates/predicates.go]: I assume you will remove this commented out code before merging :)

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (ahg-g) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Is there actually a legitimate case where metadata is not pre-computed and we still would like to proceed and compute it here? If yes, can you please document that case here. 

I hope that we can always consider this an error, and not invoke the metadata generation logic here.

**Review by ahg-g — COMMENTED** [review]
Thanks for breaking the feature into smaller PRs, very very helpful!

**Comment by ahg-g** [comment]
oh, one more thing, we should probably add an integration test for this feature, right?

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Sure, will do. (I will start to address comments of this PR after the previous PR gets stable. Otherwise, I will fall into endless "rebasing" labor work..)

> btw, is the casting to int32 actually necessary?

Yes. Golang complains if int32 and int are combined used: `invalid operation blabla: mismatched types int32 and int`.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Yes, sometime in the last commit :)

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: This always puzzled me as well. I once asked @bsalamat on this, the answer i got is that ClusterAutoscaler seems to use a non-default algorithmProvider, so they are the only consumer of the else block.

To be honest, I haven't checked the exact running path of CusterAutoscaler, and also have no knowledge on their background and usecase.

**Comment by Huang-Wei** [comment]
> oh, one more thing, we should probably add an integration test for this feature, right?

Yes, that's what I'm working on now; which will be the 6th PR of the series.

**Comment by Huang-Wei** [comment]
> Looks good to me, I guess we officially lgtm once previous PRs are merged.

Yes, just leave a LGTM comment here, which is to indicate we've got a consensus. And after the previous PR got merged, need you to do a favor to tag /lgtm.

File: pkg/scheduler/algorithm/predicates/predicates_test.go
Comment (alculquicondor) [inline pkg/scheduler/algorithm/predicates/predicates_test.go]: This is a leftover from a missing rebase, right?

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates_test.go]: Yes.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: SGTM.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Probably not. It's obvious :)

File: pkg/scheduler/algorithm/predicates/predicates_test.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates_test.go]: Yes.

File: pkg/scheduler/util/utils.go
Comment (alculquicondor) [inline pkg/scheduler/util/utils.go]: Shouldn't these be on `pkg/scheduler/testing`? Also, I don't see them being used in this PR.

Comment (Huang-Wei) [reply pkg/scheduler/util/utils.go]: Geez... My bad on rebasing, again. Thanks for the catch @alculquicondor. Updated.

**Review by alculquicondor — APPROVED** [review]
LGTM

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (ahg-g) [reply pkg/scheduler/algorithm/predicates/predicates.go]: can you please create an issue to track this? we also need an issue to track optimizing the implementation by using counters instead of pod lists in the maps.

File: pkg/scheduler/algorithm/predicates/predicates_test.go
Comment (ahg-g) [inline pkg/scheduler/algorithm/predicates/predicates_test.go]: can we add a test where an existing pod has labels that matches only one constraint of an incoming pod, basically test the case we discussed in https://github.com/kubernetes/kubernetes/pull/77760#discussion_r305609112

File: pkg/scheduler/algorithm/predicates/predicates_test.go
Comment (draveness) [inline pkg/scheduler/algorithm/predicates/predicates_test.go]: seems like a duplicate import :)

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates_test.go]: Good catch! Updated.

File: pkg/scheduler/algorithm/predicates/predicates_test.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates_test.go]: They're added in the test `TestEvenPodsSpreadPredicate_MultipleConstraints`. You can check out commit "EvenPodsSpread: update 'selfMatch' logic".

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Let me put them as TODO items in the umbrella issue for now, and will create Issues/PRs resolving them after the code gets merged.

**Comment by Huang-Wei** [comment]
@ahg-g comments addressed. PTAL.

File: pkg/scheduler/algorithm/predicates/predicates_test.go
Comment (ahg-g) [reply pkg/scheduler/algorithm/predicates/predicates_test.go]: nice!

**Comment by aermakov-zalando** [comment]
@Huang-Wei 
If I understand the code correctly, the predicate picks up information about available zones from the existing nodes. For autoscaling clusters that start at a small size, this might be a problem (in fact we've had this issue before with statefulsets and volume provisioning). If the cluster currently has nodes in just 2 failure zones, but the cloud provider supports 4, I don't see how the scheduler and cluster autoscaler would know about this. Or is this case covered somehow?

**Comment by aermakov-zalando** [comment]
It's also unclear how more complicated setups will work. For example, if I have multiple node pools, some of which support all 3 zones and some of them support only 2, wouldn't this cause the predicate and thus CA/scheduler to render the third pod targeting the latter pool unschedulable, because it'll mistakenly think there's a skew?

**Pull request merged**
